consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+
+        assert_eq!(
+            &self.db_tx.get_mainchain_tx_by_position(&tx_position).ok().flatten().expect("Database corrupted! "),
+            tx
+        );
+
+        TxMainChainIndex::new(
+            SpendablePosition::from(tx_position),
+            tx.get_outputs().len() as u32,
+        )
+        .map_err(BlockError::from)
+    }
+
+    fn connect_transactions(&mut self, block: &Block) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in block.transactions().iter() {
+            for input in tx.get_inputs() {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
+                    Some(tx_index) => tx_index.clone(),
+                    None => {
+                        let tx_index = self.calculate_indices(block, tx)?;
+                        cached_inputs.insert(tx.get_id(), tx_index.clone());
+                        tx_index
+                    }
+                };
+                if input_index >= tx_index.get_output_count() {
+                    return Err(BlockError::Unknown);
+                }
+                // Set each input as spent
+                tx_index
+                    .spend(input_index, Spender::from(tx.get_id()))
+                    .map_err(BlockError::from)?;
+                cached_inputs.insert(tx.get_id(), tx_index.clone());
+            }
+        }
+        self.store_cached_inputs(&cached_inputs)?;
+        Ok(())
+    }
+
+    fn disconnect_transactions(&mut self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in transactions.iter().rev() {
+            let inputs = tx.get_inputs();
+            for input in inputs {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
+                    Some(tx_index) => tx_index.clone(),
+                    None => self
+                        .db_tx
+                        .get_mainchain_tx_index(&tx.get_id())?
+                        .ok_or(BlockError::Unknown)?,
+                };
+
+                if input_index >= tx_index.get_output_count() {
+                    return Err(BlockError::Unknown);
+                }
+                tx_index.unspend(input_index).map_err(BlockError::from)?;
+                cached_inputs.insert(tx.get_id(), tx_index.clone());
+            }
+            // TODO: Seems, would be better to remove TxMainChainIndex for disconnected block.
+            // self.db_tx.del_mainchain_tx_index(&tx.get_id())?;
+        }
+        self.store_cached_inputs(&cached_inputs)?;
+        Ok(())
+    }
+
+    fn get_mainchain_index_by_outpoint(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
```

Comment:
It seems to me that `<blockchain_storage::Store as Transactional<'a>>::TransactionRw` deserves a type alias

Reply:
There should also be an alias that casts it to read-only. Given that this is a super-trait, that should not be a problem. Functions that only read should use the read-only trait.

But again, up to you Anton whether you create a TODO with an issue on github, or do it right now. I would like to prioritize testing everything + merging a working and tested version of this.

Reply:
I changed it to generic type. I hope it's more readable. Please, let me know if everyone is happy with it.

```rust
    fn get_mainchain_index_by_outpoint<TxRo: BlockchainStorageRead>(
        tx_db: &TxRo,  
  ```